### PR TITLE
At opt-level=0, apply only ABI-affecting attributes to functions

### PR DESCRIPTION
**Pull request by erikdesjardins**
This should provide a small perf improvement for debug builds,
and should more than cancel out the perf regression from adding noundef (https://github.com/rust-lang/rust/pull/93670#issuecomment-1038347581, #94106).

r? @nikic

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/abi.rs [hunk 1]
```diff
@@ -74,19 +80,14 @@ impl ArgAttributesExt for ArgAttributes {
             if let Some(align) = self.pointee_align {
                 llvm::LLVMRustAddAlignmentAttr(llfn, idx.as_uint(), align.bytes() as u32);
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/abi.rs:81]: The alignment can be ABI-relevant if the argument is byval. For simplicity, probably best to always add it.

**Comment by nikic** [comment]
It may be worthwhile to split our LLVM attribute APIs into ones that create an Attribute, an then another that applies a whole list of attributes to an index. What we're doing right now if really inefficient and will end up interning a lot of unnecessary AttributeSets and AttributeLists.

**Comment by erikdesjardins** [comment]
I'll work on that next, after my current batch of PRs.

Unless you're suggesting that it would make our attribute manipulation fast enough that this PR would no longer be necessary?

**Comment by nikic** [comment]
> Unless you're suggesting that it would make our attribute manipulation fast enough that this PR would no longer be necessary?

I suspect so, but I don't particularly mind the conditional attributes either.

@bors r+

**Pull request merged**